File: sdk/php/tests/e2e/rest/Account/AccountTransferTest.php
<?php

namespace Tests\e2e\rest\Account;

use JMS\Serializer\Serializer;
use JMS\Serializer\SerializerBuilder;
use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Common\Logger;
use KuCoin\UniversalSDK\Generate\Account\Transfer\FlexTransferReq;
use KuCoin\UniversalSDK\Generate\Account\Transfer\FuturesAccountTransferInReq;
use KuCoin\UniversalSDK\Generate\Account\Transfer\FuturesAccountTransferOutReq;
use KuCoin\UniversalSDK\Generate\Account\Transfer\GetFuturesAccountTransferOutLedgerReq;
use KuCoin\UniversalSDK\Generate\Account\Transfer\GetTransferQuotasReq;
use KuCoin\UniversalSDK\Generate\Account\Transfer\InnerTransferReq;
use KuCoin\UniversalSDK\Generate\Account\Transfer\SubAccountTransferReq;
use KuCoin\UniversalSDK\Generate\Account\Transfer\TransferApi;
use KuCoin\UniversalSDK\Internal\Utils\JsonSerializedHandler;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use PHPUnit\Framework\TestCase;
use Ramsey\Uuid\Uuid;

class AccountTransferTest extends TestCase
{
    /**
     * @var TransferApi $api
     */
    private $api;

    /**
     * @var Serializer $serializer
     */
    private $serializer;

    protected function setUp(): void
    {

        $this->serializer = SerializerBuilder::create()
            ->addDefaultHandlers()
            ->configureHandlers(function ($handlerRegistry) {
                $handlerRegistry->registerSubscribingHandler(
                    new JsonSerializedHandler()
                );
            })
            ->build();

        // Retrieve API secret information from environment variables
        $key = getenv('API_KEY') ?: '';
        $secret = getenv('API_SECRET') ?: '';
        $passphrase = getenv('API_PASSPHRASE') ?: '';

        // Optional: Retrieve broker secret information from environment variables; applicable for broker API only
        $brokerName = getenv('BROKER_NAME');
        $brokerKey = getenv('BROKER_KEY');
        $brokerPartner = getenv('BROKER_PARTNER');

        // Set specific options, others will fall back to default values
        $httpTransportOption = (new TransportOptionBuilder())
            ->setKeepAlive(true)
            ->setMaxConnections(10)
            ->build();

        // Create a client using the specified options
        $clientOption = (new ClientOptionBuilder())
            ->setKey($key)
            ->setSecret($secret)
            ->setPassphrase($passphrase)
            ->setBrokerName($brokerName)
            ->setBrokerKey($brokerKey)
            ->setBrokerPartner($brokerPartner)
            ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
            ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
            ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
            ->setTransportOption($httpTransportOption)
            ->build();

        $client = new DefaultClient($clientOption);
        $kucoinRestService = $client->restService();
        $this->api = $kucoinRestService->getAccountService()->getTransferApi();
    }


    /**
     * getTransferQuotas
     * Get Transfer Quotas
     * /api/v1/accounts/transferable
     */
    public function testGetTransferQuotas()
    {
        $builder = GetTransferQuotasReq::builder();
        $builder->setCurrency("USDT")->setType("MAIN")->setTag("");
        $req = $builder->build();
        $resp = $this->api->getTransferQuotas($req);
        self::assertNotNull($resp->currency);
        self::assertNotNull($resp->balance);
        self::assertNotNull($resp->available);
        self::assertNotNull($resp->holds);
        self::assertNotNull($resp->transferable);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * flexTransfer
     * Flex Transfer
     * /api/v3/accounts/universal-transfer
     */
    public function testFlexTransfer()
    {
        $builder = FlexTransferReq::builder();
        $builder->setClientOid(Uuid::uuid4()->toString())->setCurrency("USDT")->setAmount("1")->
        setFromUserId("6744227ce235b300012232d6")->setFromAccountType("MAIN")->
        setType("INTERNAL")->setToUserId("6744227ce235b300012232d6")->setToAccountType("TRADE");
        $req = $builder->build();
        $resp = $this->api->flexTransfer($req);
        self::assertNotNull($resp->orderId);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * subAccountTransfer
     * Sub-account Transfer
     * /api/v2/accounts/sub-transfer
     */
    public function testSubAccountTransfer()
    {
        $builder = SubAccountTransferReq::builder();
        $builder->setClientOid(Uuid::uuid4()->toString())->setCurrency("USDT")->setAmount("1")->setDirection("OUT")
            ->setAccountType("MAIN")->setSubAccountType("MAIN")->setSubUserId("6744227ce235b300012232d6");
        $req = $builder->build();
        $resp = $this->api->subAccountTransfer($req);
        self::assertNotNull($resp->orderId);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * innerTransfer
     * Internal Transfer
     * /api/v2/accounts/inner-transfer
     */
    public function testInnerTransfer()
    {
        $builder = InnerTransferReq::builder();
        $builder->setClientOid(Uuid::uuid4()->toString())->setCurrency("USDT")
            ->setAmount("1")->setTo("main")->setFrom("trade");
        $req = $builder->build();
        $resp = $this->api->innerTransfer($req);
        self::assertNotNull($resp->orderId);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * getFuturesAccountTransferOutLedger
     * Get Futures Account Transfer Out Ledger
     * /api/v1/transfer-list
     */
    public function testGetFuturesAccountTransferOutLedger()
    {
        $builder = GetFuturesAccountTransferOutLedgerReq::builder();
        $builder->setCurrency("USDT")->setType("MAIN");
        $req = $builder->build();
        $resp = $this->api->getFuturesAccountTransferOutLedger($req);
        self::assertNotNull($resp->currentPage);
        self::assertNotNull($resp->pageSize);
        self::assertNotNull($resp->totalNum);
        self::assertNotNull($resp->totalPage);
        foreach ($resp->items as $item) {
            self::assertNotNull($item->applyId);
            self::assertNotNull($item->currency);
            self::assertNotNull($item->recRemark);
            self::assertNotNull($item->recSystem);
            self::assertNotNull($item->status);
            self::assertNotNull($item->amount);
            self::assertNotNull($item->reason);
            self::assertNotNull($item->offset);
            self::assertNotNull($item->createdAt);
            self::assertNotNull($item->remark);
        }

        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * futuresAccountTransferOut
     * Futures Account Transfer Out
     * /api/v3/transfer-out
     */
    public function testFuturesAccountTransferOut()
    {
        $builder = FuturesAccountTransferOutReq::builder();
        $builder->setCurrency("USDT")->setAmount("1")->setRecAccountType("MAIN");
        $req = $builder->build();
        $resp = $this->api->futuresAccountTransferOut($req);
        self::assertNotNull($resp->applyId);
        self::assertNotNull($resp->bizNo);
        self::assertNotNull($resp->payAccountType);
        self::assertNotNull($resp->payTag);
        self::assertNotNull($resp->remark);
        self::assertNotNull($resp->recAccountType);
        self::assertNotNull($resp->recTag);
        self::assertNotNull($resp->recRemark);
        self::assertNotNull($resp->recSystem);
        self::assertNotNull($resp->status);
        self::assertNotNull($resp->currency);
        self::assertNotNull($resp->amount);
        self::assertNotNull($resp->fee);
        self::assertNotNull($resp->sn);
        self::assertNotNull($resp->reason);
        self::assertNotNull($resp->createdAt);
        self::assertNotNull($resp->updatedAt);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

    /**
     * futuresAccountTransferIn
     * Futures Account Transfer In
     * /api/v1/transfer-in
     */
    public function testFuturesAccountTransferIn()
    {
        $builder = FuturesAccountTransferInReq::builder();
        $builder->setCurrency("USDT")->setAmount("1")->setPayAccountType("MAIN");
        $req = $builder->build();
        $resp = $this->api->futuresAccountTransferIn($req);
        self::assertNull($resp->data);
        Logger::info($resp->jsonSerialize($this->serializer));
    }

}

File: sdk/java/src/test/java/com/kucoin/universal/sdk/test/e2e/rest/account/AccountTransferTest.java
package com.kucoin.universal.sdk.test.e2e.rest.account;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.kucoin.universal.sdk.api.DefaultKucoinClient;
import com.kucoin.universal.sdk.api.KucoinClient;
import com.kucoin.universal.sdk.generate.account.transfer.*;
import com.kucoin.universal.sdk.model.ClientOption;
import com.kucoin.universal.sdk.model.Constants;
import com.kucoin.universal.sdk.model.TransportOption;
import java.io.IOException;
import java.util.Collections;
import java.util.UUID;
import lombok.extern.slf4j.Slf4j;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

@Slf4j
public class AccountTransferTest {

  private static TransferApi api;

  public static ObjectMapper mapper = new ObjectMapper();

  @BeforeAll
  public static void setUp() {

    String key = System.getenv("API_KEY");
    String secret = System.getenv("API_SECRET");
    String passphrase = System.getenv("API_PASSPHRASE");

    TransportOption httpTransport =
        TransportOption.builder()
            .interceptors(
                Collections.singleton(
                    new Interceptor() {
                      @NotNull @Override
                      public Response intercept(@NotNull Chain chain) throws IOException {
                        Request request = chain.request();

                        System.out.println("========== Request ==========");
                        System.out.println(request.method() + " " + request.url());

                        Response response = chain.proceed(request);

                        System.out.println("========== Response ==========");
                        System.out.println("Status Code: " + response.code());
                        System.out.println("Message: " + response.message());
                        return response;
                      }
                    }))
            .build();

    ClientOption clientOpt =
        ClientOption.builder()
            .key(key)
            .secret(secret)
            .passphrase(passphrase)
            .spotEndpoint(Constants.GLOBAL_API_ENDPOINT)
            .futuresEndpoint(Constants.GLOBAL_FUTURES_API_ENDPOINT)
            .brokerEndpoint(Constants.GLOBAL_BROKER_API_ENDPOINT)
            .transportOption(httpTransport)
            .build();

    KucoinClient kucoinClient = new DefaultKucoinClient(clientOpt);
    api = kucoinClient.getRestService().getAccountService().getTransferApi();
  }

  /** getTransferQuotas Get Transfer Quotas /api/v1/accounts/transferable */
  @Test
  public void testGetTransferQuotas() throws Exception {
    GetTransferQuotasReq.GetTransferQuotasReqBuilder builder = GetTransferQuotasReq.builder();
    builder.currency("USDT").type(GetTransferQuotasReq.TypeEnum.MAIN).tag("");
    GetTransferQuotasReq req = builder.build();
    GetTransferQuotasResp resp = api.getTransferQuotas(req);
    Assertions.assertNotNull(resp.getCurrency());
    Assertions.assertNotNull(resp.getBalance());
    Assertions.assertNotNull(resp.getAvailable());
    Assertions.assertNotNull(resp.getHolds());
    Assertions.assertNotNull(resp.getTransferable());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** flexTransfer Flex Transfer /api/v3/accounts/universal-transfer */
  @Test
  public void testFlexTransfer() throws Exception {
    FlexTransferReq.FlexTransferReqBuilder builder = FlexTransferReq.builder();
    builder
        .clientOid(UUID.randomUUID().toString())
        .currency("USDT")
        .amount("1")
        .fromAccountType(FlexTransferReq.FromAccountTypeEnum.TRADE)
        .toAccountType(FlexTransferReq.ToAccountTypeEnum.MAIN)
        .type(FlexTransferReq.TypeEnum.INTERNAL);
    FlexTransferReq req = builder.build();
    FlexTransferResp resp = api.flexTransfer(req);
    Assertions.assertNotNull(resp.getOrderId());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** subAccountTransfer Sub-account Transfer /api/v2/accounts/sub-transfer */
  @Test
  public void testSubAccountTransfer() throws Exception {
    SubAccountTransferReq.SubAccountTransferReqBuilder builder = SubAccountTransferReq.builder();
    builder
        .clientOid(UUID.randomUUID().toString())
        .currency("USDT")
        .amount("1")
        .direction(SubAccountTransferReq.DirectionEnum.OUT)
        .accountType(SubAccountTransferReq.AccountTypeEnum.MAIN)
        .subAccountType(SubAccountTransferReq.SubAccountTypeEnum.MAIN)
        .subUserId("6744227ce235b300012232d6");
    SubAccountTransferReq req = builder.build();
    SubAccountTransferResp resp = api.subAccountTransfer(req);
    Assertions.assertNotNull(resp.getOrderId());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** innerTransfer Internal Transfer /api/v2/accounts/inner-transfer */
  @Test
  public void testInnerTransfer() throws Exception {
    InnerTransferReq.InnerTransferReqBuilder builder = InnerTransferReq.builder();
    builder
        .clientOid(UUID.randomUUID().toString())
        .currency("USDT")
        .amount("1")
        .to(InnerTransferReq.ToEnum.MAIN)
        .from(InnerTransferReq.FromEnum.TRADE);
    InnerTransferReq req = builder.build();
    InnerTransferResp resp = api.innerTransfer(req);
    Assertions.assertNotNull(resp.getOrderId());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /**
   * getFuturesAccountTransferOutLedger Get Futures Account Transfer Out Ledger
   * /api/v1/transfer-list
   */
  @Test
  public void testGetFuturesAccountTransferOutLedger() throws Exception {
    GetFuturesAccountTransferOutLedgerReq.GetFuturesAccountTransferOutLedgerReqBuilder builder =
        GetFuturesAccountTransferOutLedgerReq.builder();
    builder.currency("USDT").type(GetFuturesAccountTransferOutLedgerReq.TypeEnum.MAIN);
    GetFuturesAccountTransferOutLedgerReq req = builder.build();
    GetFuturesAccountTransferOutLedgerResp resp = api.getFuturesAccountTransferOutLedger(req);
    Assertions.assertNotNull(resp.getCurrentPage());
    Assertions.assertNotNull(resp.getPageSize());
    Assertions.assertNotNull(resp.getTotalNum());
    Assertions.assertNotNull(resp.getTotalPage());
    resp.getItems()
        .forEach(
            item -> {
              Assertions.assertNotNull(item.getApplyId());
              Assertions.assertNotNull(item.getCurrency());
              Assertions.assertNotNull(item.getRecRemark());
              Assertions.assertNotNull(item.getRecSystem());
              Assertions.assertNotNull(item.getStatus());
              Assertions.assertNotNull(item.getAmount());
              Assertions.assertNotNull(item.getReason());
              Assertions.assertNotNull(item.getOffset());
              Assertions.assertNotNull(item.getCreatedAt());
              Assertions.assertNotNull(item.getRemark());
            });

    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** futuresAccountTransferOut Futures Account Transfer Out /api/v3/transfer-out */
  @Test
  public void testFuturesAccountTransferOut() throws Exception {
    FuturesAccountTransferOutReq.FuturesAccountTransferOutReqBuilder builder =
        FuturesAccountTransferOutReq.builder();
    builder
        .currency("USDT")
        .amount(1.0)
        .recAccountType(FuturesAccountTransferOutReq.RecAccountTypeEnum.MAIN);
    FuturesAccountTransferOutReq req = builder.build();
    FuturesAccountTransferOutResp resp = api.futuresAccountTransferOut(req);
    Assertions.assertNotNull(resp.getApplyId());
    Assertions.assertNotNull(resp.getBizNo());
    Assertions.assertNotNull(resp.getPayAccountType());
    Assertions.assertNotNull(resp.getPayTag());
    Assertions.assertNotNull(resp.getRemark());
    Assertions.assertNotNull(resp.getRecAccountType());
    Assertions.assertNotNull(resp.getRecTag());
    Assertions.assertNotNull(resp.getRecRemark());
    Assertions.assertNotNull(resp.getRecSystem());
    Assertions.assertNotNull(resp.getStatus());
    Assertions.assertNotNull(resp.getCurrency());
    Assertions.assertNotNull(resp.getAmount());
    Assertions.assertNotNull(resp.getFee());
    Assertions.assertNotNull(resp.getSn());
    Assertions.assertNotNull(resp.getReason());
    Assertions.assertNotNull(resp.getCreatedAt());
    Assertions.assertNotNull(resp.getUpdatedAt());
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }

  /** futuresAccountTransferIn Futures Account Transfer In /api/v1/transfer-in */
  @Test
  public void testFuturesAccountTransferIn() throws Exception {
    FuturesAccountTransferInReq.FuturesAccountTransferInReqBuilder builder =
        FuturesAccountTransferInReq.builder();
    builder
        .currency("USDT")
        .amount(1.0)
        .payAccountType(FuturesAccountTransferInReq.PayAccountTypeEnum.MAIN);
    FuturesAccountTransferInReq req = builder.build();
    FuturesAccountTransferInResp resp = api.futuresAccountTransferIn(req);
    log.info("resp: {}", mapper.writeValueAsString(resp));
  }
}
